executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+                    let mut var_idx = 0;
+                    args.into_iter().for_each(|arg| {
+                        if let LookupCell::Output(v) = arg {
+                            assert!(vars[result_vars[var_idx]].replace(*v).is_none());
+                            var_idx += 1;
+                        }
+                    });
+                }
+                InterpreterAction::Assertion(e1, e2, expected_equal) => {
+                    let lhs_value = self.evaluate_expression(&mut eval_stack, &vars, e1);
+                    let rhs_value = self.evaluate_expression(&mut eval_stack, &vars, e2);
+                    if *expected_equal {
+                        assert_eq!(lhs_value, rhs_value, "Assertion failed");
+                    } else {
+                        assert_ne!(lhs_value, rhs_value, "Assertion failed");
+                    }
+                }
+            }
+        }
+    }
+
+    // evaluates an expression using the provided variable values
+    fn evaluate_expression(
+        &self,
+        stack: &mut Vec<T>,
+        vars: &[Option<T>],
+        expr: &RPNExpression<T, usize>,
+    ) -> T {
+        for elem in expr.elems.iter() {
+            match elem {
+                RPNExpressionElem::Concrete(v) => stack.push(*v),
+                RPNExpressionElem::Symbol(idx) => stack.push(vars[*idx].unwrap()),
+                RPNExpressionElem::BinaryOperation(op) => {
+                    let right = stack.pop().unwrap();
+                    let left = stack.pop().unwrap();
+                    let result = match op {
+                        BinaryOperator::Add => left + right,
+                        BinaryOperator::Sub => left - right,
+                        BinaryOperator::Mul => left * right,
+                        BinaryOperator::Div => left / right,
+                        BinaryOperator::IntegerDiv => T::from(
+                            left.to_integer().try_into_u64().unwrap()
+                                / right.to_integer().try_into_u64().unwrap(),
+                        ),
+                    };
+                    stack.push(result);
+                }
+                RPNExpressionElem::UnaryOperation(op) => {
+                    let inner = stack.pop().unwrap();
+                    let result = match op {
+                        UnaryOperator::Neg => -inner,
+                    };
+                    stack.push(result);
+                }
+                RPNExpressionElem::BitOperation(op, right) => {
+                    let left = stack.pop().unwrap();
+                    let result = match op {
+                        BitOperator::And => T::from(left.to_integer() & *right),
+                    };
+                    stack.push(result);
+                }
+            }
+        }
+        stack.pop().unwrap()
+    }
+}
+
+// the following functions come from the interface.rs file also included in the compiled jit code
```

Comment:
Does this mean it's duplicated? Any way it could be reused?

Reply:
I think we should just use `CompactDataRef`